add #![rustc_no_implicit_bounds]

davidtwco opened the pull request:
Follow-up from rust-lang/rust#137944.

Adds a new `rustc_attrs` attribute that stops rustc from adding any default bounds. Useful for tests where default bounds just add noise and make debugging harder.

After reviewing all tests with `?Sized`, these tests seem like they could probably benefit from `#![rustc_no_implicit_bounds]`.

- Skipping most of `tests/ui/unsized` as these seem to want to test `?Sized`
- Skipping tests that used `Box<T>` because it's still bound by `T: MetaSized`
- Skipping parsing or other tests that cared about `?Sized` syntactically
- Skipping tests for `derive(CoercePointee)` because this appears to check that the pointee type is relaxed with `?Sized` explicitly

r? @lcnr

davidtwco committed:
hir_analysis: add `#![rustc_no_implicit_bounds]`

Adds a new `rustc_attrs` attribute that stops rustc from adding any
default bounds. Useful for tests where default bounds just add noise and
make debugging harder.

lcnr commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:
```diff
@@ -191,6 +191,11 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
         let meta_sized_did = tcx.require_lang_item(LangItem::MetaSized, span);
         let pointee_sized_did = tcx.require_lang_item(LangItem::PointeeSized, span);
 
+        // Skip adding any default bounds if `#![rustc_no_implicit_bounds]`
+        if tcx.has_attr(CRATE_DEF_ID, sym::rustc_no_implicit_bounds) {
+            return;
+        }
```
why do this after fetching the lang items?

lcnr commented on tests/ui/cast/cast-rfc0401-vtable-kinds.rs:
```diff
@@ -1,4 +1,8 @@
 //@ run-pass
+
+#![feature(rustc_attrs)]
+#![rustc_no_implicit_bounds]
+
```
for unsize coercions/trait object coercions test, the attribute makes them a bit confusing/harder to read for me :thinking:

so maybe only limit the attribute to tests for the trait solver :thinking:

lcnr reviewed (commented):
r=me on the rustc impl. I personally feel like applying this attribute in some of the tests makes them worse :thinking: if u want I can check out your PR locally, reset the last commit and manually only readd the tests where I consider the attribute to be useful

davidtwco commented on compiler/rustc_hir_analysis/src/hir_ty_lowering/bounds.rs:
```diff
@@ -408,24 +413,21 @@ impl<'tcx> dyn HirTyLowerer<'tcx> + '_ {
     }
 
-    fn do_not_provide_default_trait_bound<'a>(
+    /// Returns `true` if default trait bound should not be added.
```
Fixed

davidtwco committed:
tests: add #![rustc_no_implicit_bounds]

After reviewing all tests with `?Sized` and discussing with lcnr, these
tests seem like they could probably benefit from
`#![rustc_no_implicit_bounds]`.

davidtwco commented:
Had a chat with @lcnr and using the attribute now in just the tests that we think are improvements.
